lib/Feature/FeatureModelWriter.cpp

```diff
@@ -0,0 +1,342 @@
+  CHECK_RC
+
+  RC = writeNumericFeatures(Writer);
+  CHECK_RC
+
+  RC = writeBooleanConstraints(Writer);
+  CHECK_RC
+
+  // TODO mixed and nonNumeric constraints when supported
+
+  RC = xmlTextWriterEndDocument(Writer); // VM
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeBinaryFeatures(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::BINARYOPTIONS);
+  CHECK_RC
+
+  for (Feature *F : Fm.features()) {
+    if (F->getKind() == Feature::FeatureKind::FK_BINARY) {
+      RC = writeFeature(Writer, *F);
+      CHECK_RC
+    }
+  }
+
+  RC = xmlTextWriterEndElement(Writer); // BINARYOPTIONS
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeNumericFeatures(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::NUMERICOPTIONS);
+  CHECK_RC
+
+  for (auto *F : Fm.features()) {
+    if (F->getKind() == Feature::FeatureKind::FK_NUMERIC) {
+      RC = writeFeature(Writer, *F);
+      CHECK_RC
+    }
+  }
+
+  RC = xmlTextWriterEndElement(Writer); // NUMERICOPTIONS
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeBooleanConstraints(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::BOOLEANCONSTRAINTS);
+  CHECK_RC
+
+  // collect alternative groups
+  std::set<std::set<Feature *>> AltGroups;
+  for (auto *F : Fm.features()) {
+    // skip empty groups
+    if (F->alternatives().begin() == F->alternatives().end()) {
+      continue;
+    }
+    std::set<Feature *> AltGroup(F->alternatives().begin(),
+                                 F->alternatives().end());
+    AltGroup.insert(F);
+    AltGroups.insert(AltGroup);
+  };
+
+  for (auto const &Group : AltGroups) {
+    RC = xmlTextWriterStartElement(Writer, XmlConstants::CONSTRAINT);
+    CHECK_RC
+
+    auto F = Group.begin();
+    RC = xmlTextWriterWriteString(Writer, BAD_CAST(*F)->getName().data());
+    CHECK_RC
+    ++F;
+    for (; F != Group.end(); ++F) {
+      RC = xmlTextWriterWriteString(Writer, BAD_CAST " | ");
+      CHECK_RC
+      RC = xmlTextWriterWriteString(Writer, BAD_CAST(*F)->getName().data());
+      CHECK_RC
+    }
+
+    RC = xmlTextWriterEndElement(Writer); // CONSTRAINT
+    CHECK_RC
+  }
+
+  // TODO write other boolean constraints if they are parsed
+
+  RC = xmlTextWriterEndElement(Writer); // BOOLEANCONSTRAINT
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeFeature(xmlTextWriterPtr Writer,
+                                        Feature &Feature1) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::CONFIGURATIONOPTION);
+  CHECK_RC
+
+  RC = xmlTextWriterWriteElement(Writer, XmlConstants::NAME,
+                                 BAD_CAST Feature1.getName().data());
+  CHECK_RC
+
+  // optional
+  RC = xmlTextWriterWriteElement(
+      Writer, XmlConstants::OPTIONAL,
+      BAD_CAST(Feature1.isOptional() ? "True" : "False"));
+  CHECK_RC
+
+  // parent
+  if (not Feature1.isRoot()) {
```

Comment:
```suggestion
  if (!Feature1.isRoot()) {
```
We should not fix `!` and `not` in one code base